### PR TITLE
README Rewrite for 2017

### PR DESCRIPTION
**Pull request by kevinmasd**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,268 +1,132 @@
...
+forseti2 (Named after the Norse god of justice) is PiE's third* iteration of a Field Control software. It has been in use from 2014-2017, with substantial changes along the way.
+
+## Deploying Driver Stations
+Driver Stations can be downloaded from PieCentral's [Releases section](https://github.com/pioneers/PieCentral/releases), or built locally using the `dawn/fieldcontrol` branch.
+
+### Installing for MacOS and GNU/Linux
+ 1. Download the `darwin-x64` or `linux-x64` release anywhere (preferably desktop)
+ 2. In Terminal, run this (but remember to update the station number and controller's address)
+ ```bash
+ echo -n "10.31.1.187" > ~/Desktop/bridge_address.txt;
```
Comment (brandonxxlee) [inline README.md]: Is this the right bridge address?

File: README.md [hunk 2]
```diff
@@ -1,268 +1,132 @@
...
+ echo -n "0" > ~/Desktop/station_number.txt;
```
Comment (brandonxxlee) [inline README.md]: Should probably note this isn't always just 0, but a number 0-3, where 0,1 are blue and 2,3 are gold

File: README.md [hunk 3]
```diff
@@ -1,268 +1,132 @@
...
+ ```
+ 3. Run Field Control Dawn
+
+### Installing for Windows
```
Comment (brandonxxlee) [inline README.md]: Do we even support windows

File: README.md (hunk 1 above)
Comment (kevinmasd) [reply README.md]: It literally says to "remember to update the station number and controller's address right above".

File: README.md (hunk 3 above)
Comment (kevinmasd) [reply README.md]: We don't know. We would need to try.

File: README.md [hunk 4]
```diff
@@ -1,268 +1,143 @@
...
+
+### [Lightweight Communications and Marshalling (LCM)](https://lcm-proj.github.io/)
+1. Follow Online instructions on how to build it on your system.
+2. Once LCM is built, move `lcm.jar` into the forseti2 directory to take advantage of `lcm-spy.sh` packet introspection.
+
+### Installing for Ubuntu
+1. Install LCM as detailed above
+2. Install other packages:
+```bash
+sudo apt-get install python-pygame;
```
Comment (nikitakit) [inline README.md]: I don't think forseti needs pygame anymore. It used to be used for xbox controllers (as part of live scoring), and maybe for UI -- but now we've standardized on doing UI in the browser

File: README.md [hunk 5]
```diff
@@ -1,268 +1,143 @@
...
+sudo apt-get install arduino;
```
Comment (nikitakit) [inline README.md]: Not sure we actually need arduino, someone would need to double-check. In any case, the arduino version used by hibike is not apt-get, so I don't think this instruction is something you want to run.

File: README.md [hunk 6]
```diff
@@ -1,268 +1,143 @@
...
+sudo pip install pyfirmata;
+```
+
+### Installing for MacOS
+1. Install LCM as detailed above
+2. Install other packages:
+```bash
+sudo pip install flask;
+sudo brew install pygame;
```
Comment (nikitakit) [inline README.md]: see above for pygame

File: README.md [hunk 7]
```diff
@@ -1,268 +1,143 @@
...
+sudo apt-get install python-pygame;
+sudo apt-get install arduino;
+sudo pip install pyfirmata;
+```
+
+### Installing for MacOS
+1. Install LCM as detailed above
+2. Install other packages:
+```bash
+sudo pip install flask;
```
Comment (nikitakit) [inline README.md]: No need for flask (it was a dependency of the old web server a few years back, which is now gone from the 2017 branch). Now we use lcm_ws_bridge, which needs tornado for the server

**Review by nikitakit — COMMENTED** [review]
a few comments regarding dependencies

File: README.md [hunk 8]
```diff
@@ -1,268 +1,143 @@
...
+forseti2 (Named after the Norse god of justice) is PiE's third* iteration of a Field Control software. It has been in use from 2014-2017, with substantial changes along the way.
+
+## Deploying Driver Stations
+Driver Stations can be downloaded from PieCentral's [Releases section](https://github.com/pioneers/PieCentral/releases), or built locally using the `dawn/fieldcontrol` branch.
+
+### Installing for MacOS and GNU/Linux
+ 1. Download the `darwin-x64` or `linux-x64` release anywhere (preferably desktop)
+ 2. In Terminal, run this (but remember to update the station number and controller's address)
+ ```bash
+ echo "Type Bridge Address:";
```
Comment (nikitakit) [inline README.md]: Quick thought: last year we had scripts like [this](https://github.com/pioneers/forseti2/blob/2017/driver_station.sh) in the repo.

If having the bash script here gets too unwieldy maybe there's some way to do a `curl -L | bash` thing pulling directly from github.

File: README.md [hunk 9]
```diff
@@ -1,268 +1,143 @@
...
+#### Outputs
+* Timer/Time: Time
+  * Publishes (every .3 secs) time elapsed in current match and stage, as well as the current stage name
+* Robot[0-3]/RobotControl: RobotControl
+  * Publishes (every .3 secs) the state of the robot (estop?, autonomous?, enabled?)
+
+### lcm_ws_bridge.py
+Bridges LCM to Websockets so JavaScript applications can access the network
+
+### GameObjectTimer.py
```
Comment (nikitakit) [inline README.md]: This file was for last year's game, but not this year's. I think I actually deleted it from the repo as part of cutting down on legacy code.

File: README.md (hunk 5 above)
Comment (kevinmasd) [reply README.md]: What about pyFirmata? Interestingly enough, why don't Macs need to run this?

**Comment by nikitakit** [comment]
Pyfirmata is used by `status_lights_driver.py`, as an interface with the status light hardware (playing a similar role to hibike). As a note the arduino requirements are just to flash the stock firmata sketch on the arduino that powers the status lights and similar hardware.

I think Mac would need this too, if you actually wanted to run the field from a mac. I guess the discrepancy in the readme is because Macs were being used to develop the non-hardware parts of field control, while the hardware parts ran on Linux.

Actually we should probably document how to run the python LCM nodes in the readme (though doesn't have to be in this PR). The story there is that `nodes.yml` (and `run_nodes.sh`) are used to simultaneously spawn all the field control nodes on the base station computer. These depend on tmux and a ruby gem called teamocil. However, you can also spawn the nodes manually or spread them across multiple computers on the same network. For example, during Scrimmage setup the base station's USB system stopped working and the status lights driver was moved to a laptop on the network instead.

**Comment by kevinmasd** [comment]
Mac Details will be added if we ever test with a Mac. Python nodes can also be explained after future runs/tests.

**Pull request merged**